### PR TITLE
KAFKA-6718: Rack Aware Replica Task Assignment

### PR DESCRIPTION
**Pull request by deegoy2**
Machines in data centre are sometimes grouped in racks. Racks provide isolation as each rack may be in a different physical location and has its own power source. When tasks are properly replicated across racks, it provides fault tolerance in that if a rack goes down, the remaining racks can continue to serve traffic.
 
This feature is already implemented at Kafka KIP-36 but we needed similar for task assignments at Kafka Streams Application layer. 
 
This features enables replica tasks to be assigned on different racks for fault-tolerance.
NUM_STANDBY_REPLICAS = x
totalTasks = x+1 (replica + active)

1. If there are no rackID provided: Cluster will behave rack-unaware
2. If same rackId is given to all the nodes: Cluster will behave rack-unaware
3. If (totalTasks >= number of racks), then Cluster will be rack aware i.e. each replica task is each assigned to a different rack. Among the available racks, it'll further preserve stickiness as well as select least loaded if not stiky. 
4. If (totalTasks < number of racks), then it will first assign tasks on different racks, further tasks will be assigned to least loaded node, cluster wide.

### CODE REVIEW

**Comment by guozhangwang** [comment]
Hi @deegoy2 , please note that whenever we are changing the encoding format of the protocol change we need to bump up the version of the protocol, and also please note that in KIP-268 we have proposed the way how to maintain backward compatibility when users upgrade to the new version.

**Comment by guozhangwang** [comment]
Also note there is another JIRA https://issues.apache.org/jira/browse/KAFKA-6642 proposed for this specific KIP. I think it is similar to your proposed solution but just listing it here:

```
Assignment of tasks to instances:

1. We assign active tasks to the instances where same task was running as active previously.
2. Active Tasks which couldn't be assigned in first step are assigned to the instances where same task was running as standby previously
3. Active tasks that still couldn't be assigned, are assigned to instances in round-robin way starting from least-loaded instance
4. Above 3 steps are same as StickyAssignor as there is only one active task for any task_id so no extra rack aware logic is required in assigning active tasks.
5. Now we have to assign standy-task, and here we assign these to instances running in racks other than the one with its active task. If we run out of racks then we can assign standby-tasks in same rack but on different instances.
6. This makes the assignment rack-aware but more of a best effort and doesn't guarantee anything. This is because we might not have capacity left in some racks or we might have more number of replicas than number of racks etc
Note: Here we are making current StickyTaskAssignor rack-aware, but doesn't change the logic drastically. For example, current assignor is only sticky for active tasks, and standby task assignment logic is not sticky as it doesn't look for where the task was assigned previously.
```

also cc @bbejeck since he has recently be working on this class.

**Comment by a-surana** [comment]
Just see if all the below checks are passing with the changes:
(a) ./gradlew checkstyleMain checkstyleTest
(b) ./gradlew findbugsMain findbugsTest -x test
(c) ./gradlew testAll

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:190]: "replication" might be confusion in this context. What about "StandbyTask assignment"

nit: Can you add this config in alphabetical order. Thx.

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:425]: Nit: we sort config based on `Importance` and within each importance group alphabetically. Please move the config to the correct place.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java:256]: nit: do we need `this` here? (we avoid `this` if not required. (wondering if `userEndPoint` needs `this` either -- if not, please remove it to clean up the code "on the side". thx)

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/ClientState.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/ClientState.java:38]: nit: add `final` -- we use `final` whenever possible.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/ClientState.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/ClientState.java:39]: Should this be `this(capacity, "");` ?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/ClientState.java
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/ClientState.java:38]: Why is the constructor public?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/ClientState.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/ClientState.java:43]: Nit: Can we break this long line into one line per parameter?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/ClientState.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/ClientState.java:46]: Nit: Can we break this long line into one line per parameter?

also, please add `final` to each parameter. Thx.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/StickyTaskAssignor.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/StickyTaskAssignor.java:123]: nit: add space after `,` -- I assume this is causing the checkstyle to fail

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/StickyTaskAssignor.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/StickyTaskAssignor.java:129]: as above

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/StickyTaskAssignor.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/StickyTaskAssignor.java:134]: nit: add `final`

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/StickyTaskAssignor.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/StickyTaskAssignor.java:152]: nit: add `final`

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/StickyTaskAssignor.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/StickyTaskAssignor.java:155]: add space after `if` -- also a checkstyle issue

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/StickyTaskAssignor.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/StickyTaskAssignor.java:160]: add space after `if` and before and after `==` -- also a checkstyle issue

**Review by mjsax — CHANGES_REQUESTED** [review]
You might need to do a `./gradlew clean checkstyleMain checkstyleTest` to retrigger checkstyle tests.

Overall the PR makes sense. It's a nice new feature. Please prepare a KIP for this. I guess we can merge this only after KIP-268 was merged completely as Guozhang pointed out already.

Can you also add some tests?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/StickyTaskAssignor.java
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/StickyTaskAssignor.java:160]: use `clientIds.isEmpty()` instead?

**Review by bbejeck — COMMENTED** [review]
Overall looks good.   Just one minor comment and we'll need to add some tests for the `StickyTaskAssignor` exercising assignments with rackIds

**Comment by chienhsingwu** [comment]
Hi, Is there a timeline to include this patch to the trunk?

**Comment by mjsax** [comment]
@chienhsingwu This PR seems to be abandoned. It targets `1.1` and not `trunk` and I would expect it need some cleanup when it's rebased to `trunk`.

@deegoy2 I will close this PR for now. Feel free to reopen if you are still interested to get it merged.

**Comment by mjsax** [comment]
Also, this feature needs a KIP as pointed out on the Jira ticket, but there was no KIP proposed yet.

**Pull request closed without merging**
